django/test/signals.py

```diff
@@ -142,7 +165,19 @@ def localize_settings_changed(*, setting, **kwargs):
 @receiver(setting_changed)
 def file_storage_changed(*, setting, **kwargs):
     if setting == "DEFAULT_FILE_STORAGE":
-        from django.core.files.storage import default_storage
+        from django.conf import DEFAULT_STORAGE_ALIAS
+        from django.core.files import storages
+        from django.core.files.storage import default_storage, get_storage_class
+        from django.test import ignore_warnings
+
+        try:
+            del storages.backends
+        except AttributeError:
+            pass
+        with ignore_warnings(category=RemovedInDjango51Warning):
+            storages._storages[DEFAULT_STORAGE_ALIAS] = get_storage_class(
+                kwargs["value"]
+            )()
```

Comment:
Do we need to ignore a warning here? `kwargs["value"]` should be a string path to the class, so is it not enough to assign it to the `STORAGES`?
```suggestion
        storages._storages[DEFAULT_STORAGE_ALIAS] = kwargs["value"]
```

Do we need to `del storages.backends`? The `storages_changed` callback should handle this for us :thinking:

Reply:
I have the same doubts to the `static_storage_changed()` receiver.

Reply:
Assigning a string path to the class to `STORAGES` is not enough. I'd have to delete `DEFAULT_STORAGE_ALIAS` from cached storages, clear `StorageHandler.backends` cached property and then assign it to the `STORAGES` which I'm not quite sure how to do in signals. 
```
del storages._storages[DEFAULT_STORAGE_ALIAS]
del storages._backends
settings.STORAGES[DEFAULT_STORAGE_ALIAS] = {"BACKEND": kwargs["value"]}  # ???
```

`storages_changed` callback is triggered `STORAGES`, `STATIC_ROOT`, `STATIC_URL` change and `file_storage_changed` is triggered on `DEFAULT_FILE_STORAGE` so it wouldn't work. I think we should handle legacy settings in separate signals to make it easier to remove.